### PR TITLE
Devicetree docs updates for 2.3, part 1

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Some updates to the reference page for the "core" APIs, and associated
follow-ups in the guides:

- centralize documentation of chosen zephyr nodes in a non-legacy
  file, provide a reference to them from the intro page in the guide
- review doxygen docstrings and correct errors for generic APIs
- add introductory text to each section in the API reference
- add missing hardware-specific reference content

Documentation for layers built on top of these is mostly left to a future "part 2",
but I do have a smattering of fixes in the guides that I noticed while I was doing this.

Full roundup in https://github.com/zephyrproject-rtos/zephyr/issues/25469.

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
>     * centralize documentation of chosen zephyr nodes in a non-legacy
>       file

I intentionally omitted some nRF9160 specific sram nodes that look a bit sketchy to me.

File: doc/guides/dts/intro.rst [hunk 1]
```diff
@@ -281,6 +282,22 @@ reg
...
+status
+    A string which describes whether the node is enabled.
+
+    The devicetree specification allows this property to have values
+    ``"okay"``, ``"disabled"``, ``"reserved"``, ``"fail"``, and ``"fail-sss"``.
+    Only the values ``"okay"`` and ``"disabled"`` are currently relevant to
+    Zephyr; use of other values currently results in undefined behavior.
+
+    A node is considered enabled if its status property is either ``"okay"`` or
+    not defined. Nodes with status ``"disabled"`` are explicitly disabled. (For
```
Comment (galak) [inline doc/guides/dts/intro.rst:294]: I wonder if 'not defined' isn't quite clear based on the previous paragraph.  Maybe 'not defined (ie does not exist in the dts)'?

Comment (pabigot) [reply doc/guides/dts/intro.rst:294]: or "not provided" or "missing" or "absent".  I see "missing" used at least once in the existin documentation:

> (howtos.rst) node's ``compatible`` property is missing, or its value has no matching

Comment (mbolivar-nordic) [reply doc/guides/dts/intro.rst:294]: changed both to 'not defined'

File: doc/reference/devicetree/index.rst [hunk 2]
```diff
@@ -9,46 +9,111 @@ scheduling. They can be used from any calling context and at file scope.
...
+some of these macros can be used even for nodes without matching bindings. This
+does not apply to macros which take cell names as arguments.
+
+.. doxygengroup:: devicetree-interrupts-prop
+   :project: Zephyr
+
+For-each macros
+===============
+
+There is currently only one generic for-each macro, :c:func:`DT_FOREACH_CHILD`,
```
Comment (galak) [inline doc/reference/devicetree/index.rst:88]: This is a little confusing.  We only have DT_FOREACH_CHILD, but then the next paragraph references DT_INST_FOREACH_STATUS_OKAY.  So saying 'only one' might be seen as confusing.

File: doc/reference/devicetree/index.rst [hunk 3]
```diff
@@ -92,20 +163,97 @@ Clocks
...
    :project: Zephyr
 
 SPI
 ===
 
+These conveniences may be used for nodes which describe either SPI controllers
+or devices, depending on the case.
+
 .. doxygengroup:: devicetree-spi
    :project: Zephyr
```
Comment (galak) [inline doc/reference/devicetree/index.rst:180]: Do we need to add DMA, PWM, fixed-partitions here?

File: include/devicetree.h [hunk 4]
```diff
@@ -321,21 +395,23 @@
 /**
  * @brief Get a property's logical length
  *
- * Here, "length" is a number of elements, which may not
- * be a size in bytes.
+ * Here, "length" is a number of elements, which may differ from the
+ * property's size in bytes.
```
Comment (galak) [inline include/devicetree.h:399]: `size in bytes` here doesn't seem right.

File: include/devicetree.h [hunk 5]
```diff
@@ -321,21 +395,23 @@
...
  *
  * These properties are handled as special cases:
  *
  * - reg property: use DT_NUM_REGS(node_id) instead
  * - interrupts property: use DT_NUM_IRQS(node_id) instead
  *
- * It is an error to use this macro with the above properties.
+ * It is an error to use this macro with the reg or interrupts properties.
  *
  * For other properties, behavior is undefined.
```
Comment (galak) [inline include/devicetree.h:416]: is it undefined or an error?

File: include/devicetree.h [hunk 6]
```diff
@@ -925,15 +1128,14 @@
...
+ * @return 1 if the node has the given status, 0 otherwise.
  */
 #define DT_NODE_HAS_STATUS(node_id, status) \
 	DT_NODE_HAS_STATUS_INTERNAL(node_id, status)
 
 /**
- * @brief Does the devicetree have an "okay" node with a compatible?
+ * @brief Does the devicetree have a status "okay" node with a compatible?
  *
  * Test for whether the devicetree has any nodes with status "okay"
```
Comment (galak) [inline include/devicetree.h:1140]: should we document "ok" and "okay" being the same here.

**Review by galak — COMMENTED** [review]
looks pretty good, a few comments/questions here are there.

File: doc/reference/devicetree/index.rst (hunk 2 above)
Comment (mbolivar-nordic) [reply doc/reference/devicetree/index.rst:88]: This is the only "generic" one.

File: include/devicetree.h (hunk 4 above)
Comment (mbolivar-nordic) [reply include/devicetree.h:399]: What is wrong about it? That looks correct to me. The sentence is saying length **may differ** from the size in bytes, i.e. they are not the same

File: include/devicetree.h (hunk 5 above)
Comment (mbolivar-nordic) [reply include/devicetree.h:416]: undefined

File: doc/reference/devicetree/index.rst (hunk 2 above)
Comment (mbolivar-nordic) [reply doc/reference/devicetree/index.rst]: Changed the wording to hopefully make this more clear.

**Review by galak — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**